dipy/align/imaffine.py

```diff
@@ -0,0 +1,715 @@
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from dipy.align import floating
+import dipy.align.vector_fields as vf
+from dipy.align.mattes import MattesBase, sample_domain_regular
+from dipy.core.optimize import Optimizer
+from dipy.align.imwarp import (get_direction_and_spacings, ScaleSpace)
+from dipy.align.scalespace import IsotropicScaleSpace
+
+
+class MattesMIMetric(MattesBase):
+    def __init__(self, nbins=32, sampling_proportion=None):
+        r""" Initializes an instance of the Mattes MI metric
+
+        This class implements the methods required by Optimizer to drive the
+        registration process by making calls to the low level methods defined
+        in MattesBase.
+
+        Parameters
+        ----------
+        nbins : int
+            the number of bins to be used for computing the intensity
+            histograms
+        sampling_proportion : int in (0,100]
+            the percentage of voxels to be used for estimating the (joint and
+            marginal) intensity histograms. If None, dense sampling is used.
```

Comment:
You are right!, I forgot to mention. I added the following note:
```
        Notes
        -----
        Since we use linear interpolation, images are not, in general,
        differentiable at exact voxel coordinates, but they are differentiable
        between voxel coordinates. When using sparse sampling, selected voxels
        are slightly moved by adding a small random displacement within one
        voxel to prevent sampling points to be located exactly at voxel
        coordinates. When using dense sampling, this random displacement is
        not applied.
```